13-multiplecomponents-lab/src/02-shoppinglist.jsx

```diff
@@ -18,11 +18,183 @@
+        super(props);
+        this.state = {
+            items: [],
+            currentItem: {}
+        }
+    }
+
+    addItem = () => {
+        let items = this.state.items;
+        let newItem = {};
+        items.push(this.state.currentItem);
+        this.setState({
+            items: items,
+            currentItem: newItem
+        });
+    }
+
+    buyItem(index){
+        console.log("update item quantity");
+        let item = this.state.items[index];
+        let items = this.state.items;
+        if (item.buyQuantity > item.quantity){
+            console.log("an error msg should be displayed, or a validator should be used");
+            item.quantity = 0
+        }else{
+            item.quantity -= item.buyQuantity;
+        }
+        item.buyQuantity = 0;
+        items[index] = item;
+        this.setState({items: items});
+
+    }
+
+    removeItem(index){
+        console.log("remove item");
+        let items = this.state.items;
+        items.splice(index,1);
+        this.setState({items: items});
+
+    }
+
+    updateQuantity(index, event){
+        let item = this.state.items[index];
+        let items = this.state.items;
+        item.buyQuantity = Number(event.target.value);
+        items[index] = item;
+        this.setState({items: items});
+
+    }
+
+    onItemChange = (evt) => {
+        let item = this.state.currentItem;
+        let itemVal = isNaN(evt.target.value) ? evt.target.value : Number(evt.target.value);
+
+        item[evt.target.name] = itemVal;
+
+        this.setState({
+            currentItem: item
+        });
+    }
+
+    render() {
+        const isTableVisible = this.state.items.length > 0;
+
+        return (<div>
+                <ShoppingItem addItem={this.addItem}
+                              name={this.state.currentItem.name}
+                              quantity={this.state.currentItem.quantity}
+                              onChange={this.onItemChange}/>
+                {isTableVisible &&
+                <div>
+                    <table>
+                        <thead>
+                        <tr>
+                            <th>Item Name</th>
+                            <th>Quantity Left</th>
+                            <th>Quantity to Buy</th>
+                            <th>Action</th>
+                        </tr>
+                        </thead>
+                        <tbody>
+                        {this.state.items.map((item, idx) => (
+                            <ShoppingListItem key={idx} item={item} index={idx} onChange={this.updateQuantity.bind(this, idx)} buyItem={this.buyItem.bind(this, idx)} removeItem={this.removeItem.bind(this, idx)} />
+                        ))}
+                        </tbody>
+                    </table>
+                </div>
+                }
+                <ShoppingItemsCount count={this.state.items.length}/>
+            </div>
+        );
+    }
+}
+
+
+export function ShoppingItem(props) {
```

Comment:
rename to ShoppingToolbar or something relevant to it's functionality.
After all it's not a shopping item